src/main/clojure/clara/rules/engine.cljc

```diff
@@ -800,15 +792,20 @@
                                   ;; presence or absence of tokens.
                                   (-> current first empty?))))
                       (mem/remove-accum-reduced! memory node join-bindings join-bindings))]
-            
-            token removed-tokens
-            ;; A nil previous result should not have been propagated before.
-            :when (some? previous-converted)
+            ;; There is nothing to do if no tokens were removed.
+            :when (seq removed-tokens)
             ;; Note that this will cause a Cartesian join between tokens and elements groups where the token
             ;; and element group share the same join bindings, but the element groups may have additional bindings
             ;; that come from their alpha nodes. Keep in mind that these element groups need elements to be created
             ;; and cannot come from initial values if they have bindings that are not shared with tokens.
-            [fact-bindings [previous previous-reduced]] previous-results]
+            [fact-bindings [previous previous-reduced]] previous-results
+            :let [;; If there were tokens before that are now removed, the value would have been accumulated already.
+                  ;; This means there is no need to check for ::not-reduced here.
+                  previous-converted (when (some? previous-reduced)
+                                       ((:convert-return-fn accumulator) previous-reduced))]
+            ;; A nil previous result should not have been propagated before.
+            :when (some? previous-converted)
```

Comment:
I'm not sure but if the tokens were empty it would be broken.  It is hard to get a scenario that does that.  I know it is definitely possible to get `right-retract` with elements that do not remove anything.  I'm not 100% sure if that is true on the `left-retract` side.  I don't really think it is worth dumping a lot of time in right now since the `seq` check is cheap enough anyways.